### PR TITLE
Update Topic partitions in provider-kafka

### PR DESCRIPTION
**Pull request by jograca**
This PR defines the Update function for Kafka Topics using UpdatePartitions

### CODE REVIEW

File: internal/controller/topic/topic.go [hunk 1]
```diff
@@ -184,9 +184,36 @@ func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
...
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+
+	p, ok := td[meta.GetExternalName(cr)]
+	if !ok || errors.Is(p.Err, kerr.UnknownTopicOrPartition) {
+		return managed.ExternalUpdate{}, nil
+	}
+
+	resp, err := c.kafkaClient.CreatePartitions(ctx, cr.Spec.ForProvider.Partitions - len(p.Partitions), meta.GetExternalName(cr))
```
[the remark this answers is not in the document]
Comment (twmb) [reply internal/controller/topic/topic.go:205]: It's not valid to decrease the number of partitions on a Kafka topic -- that operation is not supported within Kafka (and it'd likely break a lot of many client assumptions).

File: internal/controller/topic/topic.go [hunk 2]
```diff
@@ -184,9 +184,36 @@ func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
...
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+
+	p, ok := td[meta.GetExternalName(cr)]
+	if !ok || errors.Is(p.Err, kerr.UnknownTopicOrPartition) {
+		return managed.ExternalUpdate{}, nil
+	}
+
+	resp, err := c.kafkaClient.CreatePartitions(ctx, cr.Spec.ForProvider.Partitions - len(p.Partitions.Numbers()), meta.GetExternalName(cr))
```
[the remark this answers is not in the document]
Comment (twmb) [reply internal/controller/topic/topic.go]: I've added `UpdatePartitions`, which takes the final count and avoids listing internally.

File: internal/controller/topic/topic.go [hunk 3]
```diff
@@ -184,10 +184,47 @@ func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
...
+		return managed.ExternalUpdate{}, errors.Wrapf(p.Err, "cannot get topic")
+	}
+
+	l := cr.Spec.ForProvider.Partitions - len(p.Partitions)
+
+	if l < 1 {
+		return managed.ExternalUpdate{}, errors.New("cannot decrease partition count")
+	} else {
+		resp, err := c.kafkaClient.UpdatePartitions(ctx, cr.Spec.ForProvider.Partitions, meta.GetExternalName(cr))
+
```
Comment (turkenh) [inline internal/controller/topic/topic.go:212]:
```suggestion
	if l < 1 {
		return managed.ExternalUpdate{}, errors.New("cannot decrease partition count")
	} 
	resp, err := c.kafkaClient.UpdatePartitions(ctx, cr.Spec.ForProvider.Partitions, meta.GetExternalName(cr))
```
`else` could be omitted and early returns in if conditions are idiomatic in go: https://golang.org/doc/effective_go#if

File: internal/controller/topic/topic.go [hunk 4]
```diff
@@ -184,10 +184,47 @@ func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
...
+		return managed.ExternalUpdate{}, nil
+	}
+	if p.Err != nil {
+		return managed.ExternalUpdate{}, errors.Wrapf(p.Err, "cannot get topic")
+	}
+
+	l := cr.Spec.ForProvider.Partitions - len(p.Partitions)
+
+	if l < 1 {
+		return managed.ExternalUpdate{}, errors.New("cannot decrease partition count")
```
Comment (turkenh) [inline internal/controller/topic/topic.go:209]:
```suggestion
		return managed.ExternalUpdate{}, errors.Errorf("cannot decrease partition count from %d to %d", len(p.Partitions), cr.Spec.ForProvider.Partitions)
```

nit: Adding some more details with numbers could be helpful for debugging.

**Review by turkenh — APPROVED** [review]
Added two nitpicks, other than that, it looks good to me!

File: internal/controller/topic/topic.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (marshmallory) [reply internal/controller/topic/topic.go]: Error handling has been added for this issue in the most recent commit.

**Review by marshmallory — APPROVED** [review]
Added Hasan's suggestions.

File: internal/controller/topic/topic.go [hunk 5]
```diff
@@ -184,10 +184,50 @@ func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
...
+	if p.Err != nil {
+		return managed.ExternalUpdate{}, errors.Wrapf(p.Err, "cannot get topic")
+	}
+
+	l := cr.Spec.ForProvider.Partitions - len(p.Partitions)
+
+	if l < 1 {
+		return managed.ExternalUpdate{}, errors.Errorf("cannot decrease partition count from %d to %d", len(p.Partitions), cr.Spec.ForProvider.Partitions)
+	} else {
+		resp, err := c.kafkaClient.UpdatePartitions(ctx, cr.Spec.ForProvider.Partitions, meta.GetExternalName(cr))
```
Comment (twmb) [inline internal/controller/topic/topic.go:211]: I think these four lines were meant to be deleted -- they're duplicated below. As well, the formatting from line 216+ has one too many tabs.

Comment (jograca) [reply internal/controller/topic/topic.go:211]: Awesome, appreciated Travis! Will get on fixing it in the morning

Would you be interested in an overview of what we're working on with provider-kafka? Let's connect and find a time that we could get together: jon.graca@libertymutual.com

Comment (jograca) [reply internal/controller/topic/topic.go:211]: We did fix things in https://github.com/crossplane-contrib/provider-kafka/pull/10

**Pull request merged**